### ISSUE TITLE
Temporary files not offloading to S3 when used as staging environment with Snowflake with write disposition Merge

### ISSUE THREAD

**Issue by b-bokma** [dlt-hub/dlt#1907]
### dlt version

1.1.0

### Describe the problem

I have set up my dlt destination as Snowflake with an S3 bucket as staging.

In my dlt configuration I have set config items 

`dlt.config["data_writer.buffer_max_items"] = 500000
dlt.config["buffer_max_items"] = 500000
dlt.config["data_writer.file_max_bytes"] = 100000000
dlt.config["runtime.log_level"] = "DEBUG"
dlt.config["normalize.loader_file_format"] = "parquet"`

I am running dlt from Dagster on Kubernetes. 

When I notice that my memory usage is constantly expanding in the extract phase, leading to OOMKilled issues for my larger tables ( > 100 million rows), I also do not see any temporary files written anywhere, not on S3, not in my kubernetes pod. 
When extraction is finished I see the files appearing on S3, before the data is loaded to Snowflake.

This issue appears when I am using Merge as write disposition, when I use replace, my memory use is small enough.
When I do not pass a staging filesystem, the memory used by my pod also stays < 1GB.

### Expected behavior

I expect DLT to write data from memory to files, either on my staging filesystem or in the pod the job is running on, which should keep my memory footprint smaller , for all write dispositions

### Steps to reproduce

Set up a DLT Pipeline with Snowflake as destination and S3 as stage, load a larger than memory table with write disposition Merge and see the memory filling up

### Operating system

Linux, Windows

### Runtime environment

Kubernetes

### Python version

3.11

### dlt data source

dlt verified source, sql_table

### dlt destination

Filesystem & buckets, Snowflake

**Comment by rudolfix** [comment]
there's a default and quite small in memory buffer that dlt uses. some files in the pipeline working folder should be created pretty quickly. mind that we push files to staging after the extract part is fully completed.

similar problems happen when users write code that loads full source data in memory. this is sometimes simply a bug where ie. generator is converted to a list and then chunked and sent to dlt.

another reason are deployments that share memory and ephemeral disk storage like google cloud functions.

what is your data source? could  you paste the code

**Comment by b-bokma** [comment]
@rudolfix Thanks for your reply,
this is the code I use as my source, made like this so I can create a list of tables to load, all with their own definitions

`class SapHanaResource:
    """
    This class is responsible for creating a DLT resource for a SAP HANA table.
    We work with a custom SqlTableAsset model to define the table properties,
    and pass this to the default dlt sql_table function.
    Other customizations are added to methods in this class.
    """

    def __init__(self, table: SqlTableAsset, engine: Engine):
        self.table = table
        self.engine = engine
        self.limit = 1000 if settings.ENVIRONMENT == "local" else None

    def _hana_dlt_table_adapter_callback(self, sa_table: Table) -> Table:
        """Adapter function to modify the table before it is reflected."""

        for column in sa_table.columns.values():
            if isinstance(column.type, Double):
                column.type.asdecimal = False

        """Emits floats instead of numeric."""
        for column in sa_table.columns.values():
            if isinstance(column.type, Numeric):
                column.type = Float()

        return sa_table

    def _hana_query_adapter(self, sa_select: Select, sa_table: Table) -> Select:
        """Adapter function to modify the select statement before it is executed."""

        if self.limit:
            logger.info(f"Limiting query to {self.limit} rows")
            sa_select = sa_select.limit(self.limit)

        if self.table.filter:
            logger.info(f"applying filter to select statement: {self.table.filter}")
            sa_select = sa_select.where(text(self.table.filter))

        logger.info(f"created select statement: {format(sa_select.compile(compile_kwargs={'literal_binds': True}))}")

        return sa_select

    def create_resource(self):
        """
        Create a DLT resource for the SAP HANA table.
        Due to the Yield of the function, the resource is created as a generator.
        If y ou return it regularly, it will check everything on app start and repo load
        making load times bigger
        """

        @dlt.resource(
            name=self.table.target_table_name,
            primary_key=self.table.primary_keys,
            write_disposition=self.table.write_disposition
        )
        def sap_hana_resource():

            # define which write disposition is passed:
            if type(self.table.write_disposition) == dict:
                write_disposition = self.table.write_disposition.get("disposition")
            else:
                write_disposition = self.table.write_disposition

            # define incremental column
            incremental = None
            if self.table.incremental_column:
                if write_disposition in ["merge", "append"]:
                    incremental = Incremental(cursor_path=self.table.incremental_column)

            yield from sql_table(
                credentials=self.engine,
                table=self.table.source_table_name,
                schema=self.table.source_schema_name,
                chunk_size=self.table.chunk_size,
                backend=self.table.backend,
                reflection_level=self.table.reflection_level,
                table_adapter_callback=self._hana_dlt_table_adapter_callback,
                query_adapter_callback=self._hana_query_adapter,
                incremental=incremental
            )

        return sap_hana_resource


class SapHanaSource:

    def __init__(self, tables: list[SqlTableAsset]):
        self.tables = tables
        self.hana_resource = hana_resource()
        self.selected_resources = {}

    @dlt.source
    def hana(self):
        for table in self.tables:
            resource = SapHanaResource(table, engine=self.hana_resource.engine)
            yield resource.create_resource()
`

This is my table model

`class SqlTableAsset(BaseModel):
    source_schema_name: str
    source_table_name: str
    target_schema_name: str = "staging"
    target_table_name: str
    write_disposition: str | dict | None = None
    primary_keys: list[str] | None = None
    incremental_column: str | None = None
    filter: str | None = None
    # set defaults for every table
    chunk_size: int = 50000
    backend: Literal["sqlalchemy", "pyarrow", "pandas", "connectorx"] = "pyarrow"
    reflection_level: Literal["minimal", "full", "full_with_precision"] = "full_with_precision"`

and I run it from Dagster using

`@dlt_assets(
    dlt_source=sap_hana_source,
    dlt_pipeline=pipeline(
        pipeline_name="Hana_data_load",
        dataset_name=settings.SNOWFLAKE_SCHEMA,
        destination=dlt_destination.destination,
        staging=dlt_destination.stage,
        progress="log"
    ),
    name="sap_hana_assets",
    group_name="sap_hana_dlt",
    dlt_dagster_translator=DltDagsterTranslatorTablename(),
)
def dagster_hana_assets(context: AssetExecutionContext, dlt: DagsterDltResource, config: DagsterDltConfigSchema):
    yield from dlt.run(context=context)
`

**Comment by rudolfix** [comment]
thanks for the code! `sap_hana_resource` creates a resource that yield from another resource when extracting. tbh. I never tested such pattern - it will use a little bit different mechanism than typical extraction. I'll check that. Meanwhile you could try two things:

1. replace `yield from sql_table(` with `return sql_table` - this will unwrap the `sql_table` as the final resource when extracting
2. for a full fix: do not make `sap_hana_resource` a resource, instead do:
```py
table = sql_table( ... # your existing code
table.apply_hints(write_disposition=write_disposition, primary_key=primary_key)  # set write disposition and primary key on existing resource
return table
```
so dlt can extract directly from sql_table resource. btw. `sql_table` will reflect primary key from the source database. so if you have right keys there you do not need to set it

tell me if that helped with OOM. I'll write another ticket to check the case you did. we should somehow let such "ad hoc" resource to join existing extract pipeline if it exists or at least warn user.

**Comment by b-bokma** [comment]
@rudolfix 

After your comments I decided to make my code as simple as possible and include your comments with it, I ended up with

```python
def sap_hana_resource(table, engine):

    # define which write disposition is passed:
    if type(table.write_disposition) == dict:
        write_disposition = table.write_disposition.get("disposition")
    else:
        write_disposition = table.write_disposition

    # define incremental column based on write disposition
    incremental = None
    if table.incremental_column:
        if write_disposition in ["merge", "append"]:
            incremental = Incremental(cursor_path=table.incremental_column)


    def query_adapter(sa_select, sa_table):
        limit = 1000 if settings.ENVIRONMENT == "local" else None
        filter = table.filter if table.filter else None

        return _hana_query_adapter(sa_select, sa_table, limit, filter)

    logger.info(f"Found write disposition: {write_disposition} and created incremental: {incremental} from incremental column: {table.incremental_column}")

    created_table = sql_table(
        credentials=engine,
        table=table.source_table_name,
        schema=table.source_schema_name,
        chunk_size=table.chunk_size,
        backend=table.backend,
        reflection_level=table.reflection_level,
        table_adapter_callback=_hana_dlt_table_adapter_callback,
        query_adapter_callback=query_adapter,
        incremental=incremental
    )

    created_table.apply_hints(
        write_disposition=table.write_disposition,
        primary_key=table.primary_key
    )

    return created_table.with_name(table.target_table_name)


@dlt.source
def hana(tables: list[SqlTableAsset], engine: Engine):

    for table in tables:

        yield sap_hana_resource(table, engine=engine)
``` 

But i still notice the same behaviour. Data is only written to S3 after the extract phase is finished, while I would expect intermediate files to show up there as well.

I run with these settings as configuration:

- name: SQL_DEFAULT_CHUNK_SIZE
  value: "10000"
- name: DATA_WRITER__BUFFER_MAX_ITEMS
  value: "1000000"
- name: DATA_WRITER__BUFFER_MAX_BYTES
  value: "512000000"
- name: DATA_WRITER__FILE_MAX_ITEMS
  value: "5000000"
- name: DATA_WRITER__FILE_MAX_BYTES
  value: "102400000"

I also notice that when running my extract with write disposition merge, my extract is significantly slower. Do you have any ideas why that might be?

**Comment by rudolfix** [comment]
@b-bokma do you still have OOM on large tables or you see memory increasing all the time? you code is really clean now...
maybe I didn't explain this clearly: `dlt` will save extracted data in many files - according to the data writer settings above. however it will push files to staging and then to snowflake only after extract and normalize are fully completed.
data produced by a single pipeline run is loaded as a single package: as "atomic" as possible. there are no partial loads.

if your data is really large you should do a regular backfilling: run pipeline in a loop with start/end incremental values to generate many packages (if you run on Aitflow or similar - you can run in parallel) - OSS does not partition data automatically.

regarding "merge": you should not see any impact on the extract and normalize steps. the load step may be slower because we execute additional SQL queries to delete and reinsert updated data. We'll also load new data to staging dataset. This will be slower than pure append.

we generate a lot of tracing data where you'll see all the files, jobs, item counts etc. you can load tracing to snowflake after each pipeline run:
https://dlthub.com/docs/running-in-production/running#inspect-and-save-the-load-info-and-trace

**Comment by b-bokma** [comment]
Thanks @rudolfix , this really helps. I will do a full load and add the initial value for incremental in a later phase. Since I am running with dagster I need to think how to best use backfill here, but I'll work it out :-) 
I will close the issue
